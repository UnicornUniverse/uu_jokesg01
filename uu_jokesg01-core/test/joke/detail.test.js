import { Client } from "uu_appg01";
import { wait } from "uu5g05-test";
import { render, screen, userEvent } from "../tools";
import Detail from "../../src/joke/detail.js";

const jokeGet = {
  id: "joke-1",
};

const workspaceLoad = {
  data: {
    name: "The best IT jokes",
    state: "active",
  },
  sysData: {
    asidData: {
      asid: "asid-1",
    },
    awidData: {
      awid: "awid-1",
    },
    relatedObjectsMap: {
      someUrl: "http:localhost",
    },
    profileData: {
      uuIdentityProfileList: ["Authorities"],
    },
  },
  territoryData: {
    data: {
      artifact: {
        id: "artifact-1",
        uuTerritoryBaseUri: "http://localhost",
      },
    },
  },
};

const preferenceLoadFirst = {
  showAuthors: false,
};

const mockDetailView = jest.fn();
jest.mock("../../src/joke/detail-view.js", () => (props) => {
  mockDetailView(props);
  return (
    <>
      <button onClick={props.onCopyComponent}>Copy component</button>
      <button onClick={props.onOpenToNewTab}>Open to new tab</button>
    </>
  );
});

const mockCreateCopyTag = jest.fn();
const mockRedirectToPlus4UGo = jest.fn();

jest.mock("../../src/utils/utils.js", () => ({
  createCopyTag: () => mockCreateCopyTag(),
  redirectToPlus4UGo: () => mockRedirectToPlus4UGo(),
}));

function getDefaultProps() {
  return { baseUri: "https://localhost", oid: "joke-1", uu5Id: "preference-1" };
}

async function setup(props = getDefaultProps()) {
  global.URL.createObjectURL = jest.fn();
  global.URL.revokeObjectURL = jest.fn();

  mockCreateCopyTag.mockClear();
  mockRedirectToPlus4UGo.mockClear();

  Client.get.mockImplementation((uri, dtoIn) => {
    let dtoOut;

    if (uri.includes("joke/get")) dtoOut = jokeGet;
    else if (uri.includes("sys/uuAppWorkspace/load")) dtoOut = workspaceLoad;
    else if (uri.includes("preference/loadFirst")) dtoOut = preferenceLoadFirst;
    else dtoOut = null;

    return { data: dtoOut };
  });

  const user = userEvent.setup();
  const view = render(<Detail {...props} />);
  await wait();
  return { props, view, user };
}

describe(`UuJokesCore.Joke.Detail`, () => {
  it(`checks data are properly passed to DetailView`, async () => {
    await setup();

    expect(mockDetailView).toHaveBeenCalled();
    expect(mockDetailView.mock.lastCall[0]).toMatchSnapshot();
  });

  it(`checks the baseUri is properly read from context`, async () => {
    const props = { ...getDefaultProps(), baseUri: undefined };
    await setup(props);

    expect(mockDetailView.mock.lastCall[0].baseUri).not.toBeUndefined();
  });

  it("opens component to new tab", async () => {
    const { user } = await setup();

    await user.click(screen.getByRole("button", { name: "Copy component" }));

    expect(mockCreateCopyTag).toHaveBeenCalledTimes(1);
  });

  it("creates copy tag", async () => {
    const { user } = await setup();

    await user.click(screen.getByRole("button", { name: "Open to new tab" }));

    expect(mockRedirectToPlus4UGo).toHaveBeenCalledTimes(1);
  });
});
