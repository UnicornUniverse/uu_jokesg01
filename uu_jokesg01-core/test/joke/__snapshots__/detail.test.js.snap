// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UuJokesCore.Joke.Detail checks data are properly passed to DetailView 1`] = `
Object {
  "aspectRatio": undefined,
  "awscDataObject": Object {
    "data": Object {
      "data": Object {
        "artifact": Object {
          "id": "artifact-1",
          "uuTerritoryBaseUri": "http://localhost",
        },
      },
    },
  },
  "borderRadius": undefined,
  "card": "full",
  "className": undefined,
  "colorScheme": "building",
  "disabled": undefined,
  "elementAttrs": undefined,
  "elementRef": undefined,
  "fullTextSearchPriority": undefined,
  "height": undefined,
  "hidden": undefined,
  "id": undefined,
  "identificationType": undefined,
  "isHome": false,
  "jokeDataObject": Object {
    "data": Object {
      "id": "joke-1",
    },
    "errorData": null,
    "handlerMap": Object {
      "addRating": [Function],
      "load": [Function],
      "setData": [Function],
      "update": [Function],
      "updateVisibility": [Function],
    },
    "pendingData": null,
    "state": "ready",
  },
  "jokesDataObject": Object {
    "data": Object {
      "name": "The best IT jokes",
      "state": "active",
    },
    "errorData": null,
    "handlerMap": Object {
      "init": [Function],
      "load": [Function],
      "loadBasicData": [Function],
      "setState": [Function],
      "update": [Function],
    },
    "pendingData": null,
    "state": "ready",
  },
  "jokesPermission": Object {
    "category": Object {
      "canCreate": [Function],
      "canManage": [Function],
    },
    "joke": Object {
      "canAddRating": [Function],
      "canCreate": [Function],
      "canManage": [Function],
      "canUpdateVisibility": [Function],
    },
    "jokes": Object {
      "canInit": [Function],
      "canSetState": [Function],
      "canUpdate": [Function],
    },
  },
  "level": undefined,
  "nestingLevel": undefined,
  "noPrint": undefined,
  "onCopyComponent": [Function],
  "onOpenToNewTab": [Function],
  "preferenceDataObject": Object {
    "data": Object {
      "disableUserPreference": true,
      "showAuthor": true,
      "showCategories": true,
      "showCreationTime": true,
    },
    "errorData": null,
    "handlerMap": Object {
      "load": [Function],
      "save": [Function],
      "setData": [Function],
    },
    "pendingData": null,
    "state": "ready",
  },
  "showDelete": false,
  "significance": "common",
  "size": "m",
  "style": undefined,
  "width": undefined,
}
`;
